### DISCUSSION TITLE
ArrayDiffusion and ArrayDGDiffusion

### DISCUSSION THREAD

**Discussion by dingqiushi** [idaholab/moose#21857] [Q&A General] [answered]
Why is it necessary to distinguish between diagonal and non-diagonal elements when calculating Jacobi matrices in `ArrayDiffusion.C`, and why is it necessary to use two functions,` computeQpJacobian()` and `computeQpOffDiagJacobian()`? Why is there no `computeQpOffDiagJacobian() ` in ArrayDGDiffusion.C, only `computeQpJacobian()` ?

**Comment by GiudGiud** [comment, marked as answer]
Hello

when there is no off-diagonal coefficients, we dont implement this routine. Are we missing a coefficient for that kernel?

It s necessary to have the contributions split between diagonal and off-diagonal because for preconditioning we sometimes only use the diagonal of the Jacobian instead of the full Jacobian. Having this separate routine allows us to form this preconditioning. 

Guillaume
